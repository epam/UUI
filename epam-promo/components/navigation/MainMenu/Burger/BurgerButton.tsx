import { withMods, IHasIcon, IDropdownToggler } from '@epam/uui-core';
import { Button, ButtonProps } from '@epam/uui-components';
import { ReactComponent as SvgTriangle } from '../../../../icons/triangle.svg';
import css from './BurgerButton.scss';

export interface BurgerButtonMods extends IHasIcon, IDropdownToggler {
    type?: 'primary' | 'secondary';
    indentLevel?: number;
}

export const BurgerButton = withMods<ButtonProps, BurgerButtonMods>(
    Button,
    props => [
        css.root,
        css['button-' + (props.type || 'primary')],
        css['indent-' + (props.indentLevel || 0)],
        props.isDropdown && css.dropdown,
        props.icon && css.hasIcon,
    ],
    () => ({ dropdownIcon: SvgTriangle, dropdownIconPosition: 'left', role: 'menuitem' })
);
