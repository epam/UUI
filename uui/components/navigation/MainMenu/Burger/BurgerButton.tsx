import React from 'react';
import { IHasIcon, IDropdownToggler, ICanBeActive, useIsActive, uuiMod } from '@epam/uui-core';
import { Button, ButtonProps } from '@epam/uui-components';
import { ReactComponent as SvgTriangle } from '@epam/assets/icons/navigation-arrow_down-outline.svg';
import css from './BurgerButton.module.scss';

interface BurgerButtonMods extends IHasIcon, IDropdownToggler {
    /*
    * Defines component type.
    */
    type?: 'primary' | 'secondary';
    /*
    * Defines components' indent.
    */
    indentLevel?: number;
}

export type BurgerButtonProps = ButtonProps & BurgerButtonMods & ICanBeActive;

function applyBurgerButtonMods(props: BurgerButtonProps) {
    return [
        css.root,
        'uui-main_menu-burger-button',
        css['button-' + (props.type || 'primary')],
        css['indent-' + (props.indentLevel || 0)],
        props.isActive && uuiMod.active,
        props.isDropdown && css.dropdown,
        props.icon && css.hasIcon,
    ];
}

export const BurgerButton = React.forwardRef<HTMLButtonElement | HTMLAnchorElement, BurgerButtonProps>((props, ref) => {
    const { isActive } = useIsActive({
        isLinkActive: props.isLinkActive,
        link: props.link,
        isActive: props.isActive,
    });

    return (
        <Button
            { ...props }
            ref={ ref }
            cx={ applyBurgerButtonMods({ ...props, isActive }) }
            rawProps={ {
                ...props.rawProps,
                role: 'menuitem',
            } }
            dropdownIcon={ SvgTriangle }
        />
    );
});
