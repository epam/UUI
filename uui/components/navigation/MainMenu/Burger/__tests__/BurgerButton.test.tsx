import React from 'react';
import { BurgerButton } from '../BurgerButton';
import { renderSnapshotWithContextAsync } from '@epam/uui-test-utils';
import { ReactComponent as CalendarIcon } from '@epam/assets/icons/action-calendar-fill.svg';

describe('BurgerButton', () => {
    it('should be rendered correctly with minimum props', async () => {
        const tree = await renderSnapshotWithContextAsync(<BurgerButton />);
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(
            <BurgerButton
                caption="Test button"
                icon={ CalendarIcon }
                href="#"
                target="_blank"
                type="secondary"
                isDropdown
                isOpen={ false }
                isActive={ true }
            />,
        );
        expect(tree).toMatchSnapshot();
    });
});
