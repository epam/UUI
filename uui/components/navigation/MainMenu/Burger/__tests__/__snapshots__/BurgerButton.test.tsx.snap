// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BurgerButton should be rendered correctly 1`] = `
<DocumentFragment>
  <a
    aria-expanded="false"
    aria-haspopup="true"
    class="uui-button-box uui-enabled -clickable uui-anchor container root uui-main_menu-burger-button button-secondary indent-0 uui-active dropdown hasIcon"
    href="#"
    rel="noopener noreferrer"
    role="menuitem"
    tabindex="0"
    target="_blank"
  >
    <div
      class="container uui-icon uui-enabled"
    >
      <svg
        class=""
      />
    </div>
    <div
      class="uui-caption"
    >
      Test button
    </div>
    <div
      class="container uui-icon uui-enabled"
    >
      <svg
        class=""
      />
    </div>
  </a>
</DocumentFragment>
`;

exports[`BurgerButton should be rendered correctly with minimum props 1`] = `
<DocumentFragment>
  <button
    class="uui-button-box uui-enabled container root uui-main_menu-burger-button button-primary indent-0"
    role="menuitem"
    tabindex="-1"
    type="button"
  />
</DocumentFragment>
`;
